form/data_mappers.rst

```diff
@@ -0,0 +1,197 @@
+            }
+
+            // invalid data type, this message will not be shown to the user (see below)
+            if (!$data instanceof Color) {
+                throw new UnexpectedTypeException($data, Color::class);
+            }
+
+            $forms = iterator_to_array($forms);
+
+            // set form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        public function mapFormsToData($forms, &$data)
+        {
+            $forms = iterator_to_array($forms);
+
+            // get form field values
+            $red = $forms['red']->getData();
+            $green = $forms['green']->getData();
+            $blue = $forms['blue']->getData();
+
+            $data = new Color($red, $green, $blue);
+        }
+    }
+
+All data mappers have to implement :class:`Symfony\\Component\\Form\\DataMapperInterface`.
+This interface contains two methods:
+
+``mapDataToForms()``
+    Recieves the current view data and a :phpclass:`RecursiveIteratorIterator`
+    with all form fields. It's task is to map the view data to the form fields.
+    The view data might be ``null`` if there is no data yet (e.g. in a "Create
+    new color" form);
+
+``mapFormsToData($forms, &$data)``
+    Maps the form field values to the ``$data`` property, which is changed by
+    reference. It recieves the form fields in a :phpclass:`RecursiveIteratorIterator`
+    and the view data.
+
+When an errors occurs, a
+:class:`Symfony\\Component\\Form\\Exception\\TransformationFailedException` or
+one of its subclasses should be thrown. The exception message won't be shown to
+the user. Instead, the ``invalid_message`` option will be used.
+
+.. caution::
+
+    The data passed to the mapper is *not yet validated*. This means that your
+    objects should allow being created in an invalid state in order to produce
+    user-friendly errors in the form.
+
+Using the Mapper
+----------------
+
+You're ready to use the data mapper for the ``ColorType`` form. Use the
+:method:`Symfony\\Component\\Form\\FormBuilderInterface::setDataMapper`
+method to configure the data mapper::
+
+    // src/AppBundle/Form/ColorType.php
+    namespace AppBundle\Form;
+
+    use AppBundle\Form\DataMapper\ColorMapper;
+
+    // ...
+    class ColorType extends AbstractType
+    {
+        public function buildForm(FormBuilderInterface $builder, array $options)
+        {
+            $builder
+                ->add('red', 'integer')
+                ->add('green', 'integer')
+                ->add('blue', 'integer')
+
+                ->setDataMapper(new ColorMapper())
+            ;
+        }
+
+        public function configureOptions(OptionsResolver $resolver)
+        {
+            $resolver->setDefaults(array(
+                // when creating a new color, the initial data should be null
+                'empty_data' => null,
```

Comment:
This is the default unless `data_class` is set or if some data prepopulates the form

Reply:
Afaics, the default it `''` and not `null`. I think I might added this because of that